### ISSUE TITLE
service failed to clean up consumer for deleted trigger

### ISSUE THREAD

**Issue by bjustin-ibm** [openwhisk/openwhisk-package-kafka#111]
I observed an incident where the service became entirely unresponsive (HTTP requests not handled, messages not consumed, triggers not fired). The last meaningful log messages looked like this:

```
2017-01-08 02:16:45,225 [INFO]: [<TRIGGER>] Cleaning up consumer
%3|1483841836.386|FAIL|rdkafka#consumer-339| sasl_ssl:///<BROKER>/1: 
Receive failed: No error
%3|1483841836.386|ERROR|rdkafka#consumer-339| sasl_ssl:///<BROKER>/1:
 Receive failed: No error
%3|1483841862.127|FAIL|rdkafka#consumer-1| sasl_ssl://<BROKER>/0: Re
ceive failed: No error
```

Incidentally, `1483841836.386` is approximately `2017-01-08 02:16` so these messages are basically consecutive.

The messages from `rdkafka` repeat indefinitely and the expected log message `Successfully closed KafkaConsumer` is never seen.

**Comment by bjustin-ibm** [comment]
**Theory:** There is a bug in the `close()` function that cannot handle being called simultaneously from separate threads.
**Conclusion:** The logs from the incidents where this problem has occurred do not show multiple triggers being deleted simultaneously.

**Comment by bjustin-ibm** [comment]
**Theory:** The machine was under heavy CPU or memory load at the time this incident occurred, contributing to the failure to properly close the consumer in a timely fashion.
**Conclusion:** Grafana logs indicate that the container was under nominal CPU (< 15%) and memory load (< 3%) before, during, and after the incident.

**Comment by dubeejw** [comment]
**Theory: ** Kafka brokers were unresponsive when the `close()` method was invoked.
**Conclusion: ** Using fake brokers resulted in the ability to successfully create and delete a trigger.

**Comment by dubeejw** [comment]
**Theory: ** the call to delete a trigger from the database is not thread-safe.
**Conclusion: ** TBD

**Comment by dubeejw** [comment]
**Theory:** A race condition exists when a user tries to delete a consumer from the provider.
**Conclusion:** TBD

**Comment by bjustin-ibm** [comment]
**Theory:** A defect in the underlying kafka library causes an indefinite hang while calling `consumer.close()`
**Evidence:** The logs indicate that the application becomes unresponsive when calling either `unsubscribe()` or `close()` on the consumer. Further experiments narrow it down to the `close()` call by showing that all threads block while calling `consumer.close()` (see [this gist](https://gist.github.com/bjustin-ibm/5cacbefc969bc2fcfccb7b81cce3f450) demonstrating the issue)
- [KafkaConsumer close is blocking forever - edenhill/librdkafka#1002](https://github.com/edenhill/librdkafka/issues/1002)
- [close() blocks all threads - confluentinc/confluent-kafka-python#107]( https://github.com/confluentinc/confluent-kafka-python/issues/107) (opened by me)

**Comment by bjustin-ibm** [comment]
Both fixes are scheduled to be released in the respective 0.9.4 versions of these projects. I don't have a release date, but I do see that they have a 0.9.4 RC2 available.
